File: 2023/07.06/송기훈/최솟값과최댓값.java
package baekjoon.gold.one;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

// 세그먼트 트리 개념 https://yoongrammer.tistory.com/103
public class 최솟값과최댓값 {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int N, M;
    static int[] numbers, minTree, maxTree;

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        numbers = new int[N + 1];
        for (int i = 1; i < N + 1; i++) {
            numbers[i] = Integer.parseInt(br.readLine());
        }

        minTree = new int[N * 4];
        maxTree = new int[N * 4];

        initMin(1, N, 1);
        initMax(1, N, 1);

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            System.out.println(min(1, N, 1, a, b) + " " + max(1, N, 1, a, b));
        }
    }

    static int initMin(int start, int end, int node) {
        if (start == end) {
            return minTree[node] = numbers[start];
        }
        int mid = (start + end) / 2;

        return minTree[node] = Math.min(
                initMin(start, mid, node * 2),
                initMin(mid + 1, end, node * 2 + 1)
        );
    }

    static int initMax(int start, int end, int node) {
        if (start == end) {
            return maxTree[node] = numbers[start];
        }
        int mid = (start + end) / 2;

        return maxTree[node] = Math.max(
                initMax(start, mid, node * 2),
                initMax(mid + 1, end, node * 2 + 1)
        );
    }

    public static int min(int start, int end, int node, int left, int right) {
        if (left > end || right < start) {
            return Integer.MAX_VALUE;
        }
        if (start == end) {
            return numbers[start];
        }
        if (left <= start && end <= right) {
            return minTree[node];
        }
        int mid = (start + end) / 2;
        return Math.min(
                min(start, mid, node * 2, left, right),
                min(mid + 1, end, node * 2 + 1, left, right)
        );
    }

    public static int max(int start, int end, int node, int left, int right) {
        if (left > end || right < start) {
            return 0;
        }
        if (start == end) {
            return numbers[start];
        }
        if (left <= start && end <= right) {
            return maxTree[node];
        }
        int mid = (start + end) / 2;
        return Math.max(
                max(start, mid, node * 2, left, right),
                max(mid + 1, end, node * 2 + 1, left, right)
        );
    }

}

File: 2023/07.06/조수연/최솟값과_최댓값.kt
import java.awt.Point
import java.io.*
import java.util.*
import kotlin.math.*

lateinit var numArr: IntArray
lateinit var tree: Array<Point>

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val (n, m) = readLine().trim().split(" ").map { it.toInt() }

    numArr = IntArray(n + 1)
    tree = Array(n * 4) { Point() }

    repeat(n) { numArr[it + 1] = readLine().toInt() }

    init(1, n, 1)

    val sb = StringBuilder()

    repeat(m) {
        val token = StringTokenizer(readLine())
        val l = token.nextToken().toInt()
        val r = token.nextToken().toInt()

        val result = find(1,n,l,r,1)

        sb.append("${result.y} ${result.x}\n")
    }

    bw.write(sb.toString())

    bw.flush()
    bw.close()
}

// 구간 합 트리 생성
fun init(start: Int, end: Int, node: Int): Point {

    if (start == end) {
        tree[node] = Point(numArr[start], numArr[end])
        return tree[node]
    }

    val mid = (start + end) / 2
    val left = init(start, mid, node * 2)
    val right = init(mid+1, end, node * 2 + 1)

    tree[node] = Point(max(left.x, right.x), min(left.y, right.y))

    return tree[node]
}

fun find(start: Int, end: Int, l : Int, r : Int, node : Int) : Point{
    // 범위 밖에 있는 경우
    if (end < l || start > r) return Point(Integer.MIN_VALUE,Integer.MAX_VALUE)
    // 범위 안에 있는 경우
    if (l <= start && end <= r) return tree[node]

    // 그렇지 않으면 left, right 두 부분으로 나누어서 찾기
    val mid = (start + end) / 2
    val left = find(start, mid,l,r, node * 2)
    val right = find(mid+1, end,l,r, node * 2 + 1)
    return Point(max(left.x,right.x),min(left.y,right.y))
}